nre/bpf/nodefw/nodefw-ebpf/src/main.rs

```diff
@@ -0,0 +1,128 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+#![no_std]
+#![no_main]
+
+use aya_bpf::{
+    bindings::xdp_action,
+    helpers::gen::bpf_ktime_get_ns,
+    macros::{map, xdp},
+    maps::HashMap,
+    programs::XdpContext,
+};
+use aya_log_ebpf::info;
+use core::mem;
+// TODO see if this is preferred over ptr_at
+// use memoffset::offset_of;
+use network_types::{
+    eth::{EthHdr, EtherType},
+    ip::{IpProto, Ipv4Hdr, Ipv6Hdr},
+    tcp::TcpHdr,
+    udp::UdpHdr,
+};
+use nodefw_common::Rule;
+
+const MAX_BLOCKLIST_SIZE: u32 = 1024;
+
+// the key is an ipv4 or ipv6 octet value expressed as an array.
+#[map]
+static BLOCKLIST: HashMap<[u8; 16usize], Rule> = HashMap::with_max_entries(MAX_BLOCKLIST_SIZE, 0);
+
+fn block_ip(ctx: &XdpContext, address: [u8; 16usize], dest_port: u16) -> bool {
+    unsafe {
+        // TODO find a way to check map len, if possible
+        // if BLOCKLIST.len() == MAX_BLOCKLIST_SIZE {
+        //     return true;
+        // }
+        if let Some(rule) = BLOCKLIST.get(&address) {
+            // TODO inspect ttl and handle that case
+            if rule.port == dest_port {
+                return true;
+            }
+        }
+        false
+    }
+}
+
+#[xdp]
+pub fn nodefw(ctx: XdpContext) -> u32 {
+    match try_nodefw(ctx) {
+        Ok(ret) => ret,
+        Err(_) => xdp_action::XDP_ABORTED,
+    }
+}
+
+#[inline(always)]
+fn ptr_at<T>(ctx: &XdpContext, offset: usize) -> Result<*const T, ()> {
+    let start = ctx.data();
+    let end = ctx.data_end();
+    let len = mem::size_of::<T>();
+
+    if start + offset + len > end {
+        return Err(());
+    }
+
+    Ok((start + offset) as *const T)
+}
+
+fn get_dest_port(ctx: &XdpContext, af: IpProto, proto: IpProto) -> Result<u16, ()> {
+    let offset = match af {
+        IpProto::Ipv4 => Ipv4Hdr::LEN,
+        IpProto::Ipv6 => Ipv6Hdr::LEN,
+        _ => {
+            info!(ctx, "invalid address family!");
```

Comment:
```suggestion
            error!(ctx, "invalid address family!");
```

Reply:
gah! my bad.  will change.  Eventually, we won't log at all here, but for testing, i'm going to leave it in.  logs will be conveyed through a perf array to the userspace for logging there (eventually)